### PR TITLE
Add validation endpoint to Chef Server.

### PR DESCRIPTION
**Pull request by markan**
Add validation endpoint to Chef Server.

We sometimes have the need to be able to determine if a request signed
with a client key is valid.  This adds an endpoint
/organizations/ORGNAME/validate/PATH that accepts a forwarded post,
and checks the signing as if it were sent to the PATH resource (PATH
can be any resource string). If the request validates, it returns 200,
and some potentially useful data. If it doesn't it returns 401 like
any other request with a bad signature.
- [x] Verify it works with the nginx code
- [x] Add RAML docs
- [x] Wait for chef server release to finish.
- [x] Rebase/Squash

### CODE REVIEW

File: src/oc_erchef/apps/oc_chef_wm/src/oc_chef_wm_base.erl [hunk 1]
```diff
@@ -987,7 +999,8 @@ verify_request_signature(Req,
             PublicKey = select_user_or_webui_key(Req, Requestors),
             Body = body_or_default(Req, <<>>),
             HTTPMethod = method_as_binary(Req),
-            Path = iolist_to_binary(wrq:path(Req)),
+            Path = Extractor(path, Req, State),
+            io:format("reqsig: ~p~n", [Path]),
```
Comment (stevendanna) [inline src/oc_erchef/apps/oc_chef_wm/src/oc_chef_wm_base.erl]: Let's use `lager` if we want to keep this in.

**Comment by stevendanna** [comment]
Overall I like this idea as I've wanted this feature for a while so that add-ons don't need to do the "go to the principals endpoint, then do the auth yourself" dance.  A few high level questions and comments:

- Do we have a planned use for this currently or is this just scratching an itch?
- If we have a planned use, did we consider whether to implement this RFC instead: https://github.com/chef/chef-rfc/blob/master/rfc038-token-authentication.md

File: src/oc_erchef/apps/oc_chef_wm/src/chef_wm_validate.erl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+process_post(Req, State) ->
+    {Result, Req1, State1} = to_json(Req, State),
+    case Result of
+        {halt, _} -> {Result, Req1, State1};
+        _ -> {true, wrq:set_resp_body(Result, Req1), State1}
+    end.
+
+to_json(Req, #base_state{requestor = #chef_requestor{name = Name, type = Type}} = State) ->
+    %% return org, client/user auth'd as.
```
Comment (stevendanna) [inline src/oc_erchef/apps/oc_chef_wm/src/chef_wm_validate.erl]: Doesn't appear that we are returning the org.

File: src/oc_erchef/apps/oc_chef_wm/src/oc_chef_wm_base.erl [hunk 3]
```diff
@@ -414,10 +416,18 @@ check_permission(Perm, AuthzObjectType, AuthzId, Req, #base_state{requestor_id=R
...
+
+is_authorized(Req, State) ->
+    is_authorized(Req, State, fun authorization_data_extractor/3).
+
 %% part of being authorized is being a member of the org; otherwise we
 %% fail out early.
-is_authorized(Req, State) ->
-    case verify_request_signature(Req, State) of
+is_authorized(Req, State, Extractor) ->
+    case verify_request_signature(Req, State, Extractor) of
```
Comment (stevendanna) [inline src/oc_erchef/apps/oc_chef_wm/src/oc_chef_wm_base.erl]: I don't see a verify_request_signature/3 function yet

**Comment by stevendanna** [comment]
Passing Wilson build: http://wilson.ci.chef.co/job/chef-server-12-trigger-ad_hoc/288/downstreambuildview/

File: src/oc_erchef/apps/oc_chef_wm/src/chef_wm_validate.erl [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    {true, Req2, State1}.
+
+to_json(Req, State) ->
+    make_body(Req, State).
+
+from_json(Req, State) ->
+    make_and_set_body(Req, State).
+
+validate_extractor(path, Req, _State) ->
+    %% wrq:disp_path/1 returns the path portion matched by "*", ie.
```
Comment (stevendanna) [inline src/oc_erchef/apps/oc_chef_wm/src/chef_wm_validate.erl]: 👍 Nice comment, had to look this up the first time I read this.

**Comment by stevendanna** [comment]
As a follow up, I just remembered that we now have raml docs for API endpoints, perhaps we should add some docs for this endpoint once we are done with the nginx-side feature?

**Comment by srenatus** [comment]
@stevendanna It would make sense, yes. However, we should probably discuss if this endpoint is ever going to be outside facing, or if it's just an internal thing to nginx offloading signed header authn for other services... So we'd maybe rather just document the external API endpoint that is going to be protected by this? 🤔

**Comment by srenatus** [comment]
LGTM?

**Comment by srenatus** [comment]
Rebased and squashed... ✅

**Pull request merged**